src/Domain/DonationRepository.php

```diff
@@ -345,6 +345,43 @@ public function findWithExpiredMatching(\DateTimeImmutable $now): array
             ->getResult();
     }
 
+    /**
+     * @return Donation[]   Donations which, when considered in isolation, could have some or all of their match
+     *                      funds swapped with higher priority matching (e.g. swapping out champion funds and
+     *                      swapping in pledges). The caller shouldn't assume that *all* donations may be fully
+     *                      swapped; typically we will choose to swap earlier-collected donations first, and it may
+     *                      be that priority funds are used up before we get to the end of the list.
+     */
+    public function findWithMatchingWhichCouldBeReplacedWithHigherPriorityAllocation(\DateTimeImmutable $cutoff): array
+    {
+        $qb = $this->getEntityManager()->createQueryBuilder()
+            ->select('d')
+            ->from(Donation::class, 'd')
+            // Only select donations with 1+ FWs (i.e. some matching).
+            ->innerJoin('d.fundingWithdrawals', 'fw')
+            ->innerJoin('fw.campaignFunding', 'donationCf')
+            ->innerJoin('d.campaign', 'c')
+            // Join CampaignFundings allocated to campaign `c` with some amount available and a lower allocationOrder
+            // than the funding of `fw`.
+            ->innerJoin(
+                'c.campaignFundings',
+                'availableCf',
+                'WITH',
+                'availableCf.amountAvailable > 0 AND availableCf.allocationOrder < donationCf.allocationOrder'
+            )
+            ->where('d.donationStatus IN (:collectedStatuses)')
+            ->andWhere('d.collectedAt > :checkAfter')
+            ->groupBy('d.id')
+            ->setParameter('collectedStatuses', DonationStatus::SUCCESS_STATUSES)
+            ->setParameter('checkAfter', $cutoff)
+        ;
```

Comment:
This is a relatively complex query. IMHO not worth unit testing but maybe worth integration testing.

Also I think the query builder code is harder to read than it would be if this was done as SQL or DQL code - maybe not worth changing now but IMHO the query builder is very rarely worthwhile, unless the structure of the query has to be dynamically generated. As Ocramius says at https://www.reddit.com/r/PHP/comments/a58yt9/comment/eblvfmd/ (via https://github.com/symfony/demo/issues/917 )

Reply:
If you want to convert to DQL it should be easy using `echo $qb->getDQL();` and then pasting it in. I tried it and converts to the code shown below (I just added line breaks) but also gives the error as shown at https://github.com/thebiggive/matchbot/pull/737#issuecomment-1858903937 , both before and after conversion. Also not sure what the last line means.


```DQL
            SELECT d FROM MatchBot\Domain\Donation d 
            INNER JOIN d.fundingWithdrawals fw 
            INNER JOIN fw.campaignFunding donationCf 
            INNER JOIN d.campaign c 
            INNER JOIN c.campaignFundings availableCf WITH availableCf.amountAvailable > 0 AND availableCf.allocationOrder < donationCf.allocationOrder
            WHERE d.donationStatus IN (:collectedStatuses) AND
            d.collectedAt > :checkAfter 
            GROUP BY d.idChecked
            0 donations and redistributed matching for 0
```